File: core/toolloop/truncate_test.go
package toolloop

import (
	"testing"

	"github.com/ZacharyZcR/NLUI/core/llm"
)

func msgs(roles ...string) []llm.Message {
	out := make([]llm.Message, len(roles))
	for i, r := range roles {
		out[i] = llm.Message{Role: r, Content: "x"}
	}
	return out
}

func TestTruncateDisabledWhenZero(t *testing.T) {
	m := msgs("system", "user", "assistant")
	got := truncateMessages(m, 0)
	if len(got) != 3 {
		t.Fatalf("expected 3, got %d", len(got))
	}
}

func TestTruncateKeepsSystem(t *testing.T) {
	m := []llm.Message{
		{Role: "system", Content: "system prompt here"},
		{Role: "user", Content: "this is a user message with enough chars"},
		{Role: "assistant", Content: "this is an assistant reply with enough chars"},
		{Role: "user", Content: "another user message with enough content"},
		{Role: "assistant", Content: "another assistant reply with enough text"},
	}
	// Budget so tight only system fits (system ~4 tokens, budget 5 leaves ~1 for rest)
	got := truncateMessages(m, 5)
	if len(got) != 1 || got[0].Role != "system" {
		t.Fatalf("expected only system, got %d messages", len(got))
	}
}

func TestTruncatePreservesRecent(t *testing.T) {
	m := []llm.Message{
		{Role: "system", Content: "s"},
		{Role: "user", Content: "old message that is quite long to eat budget"},
		{Role: "assistant", Content: "old reply that is also long enough to matter"},
		{Role: "user", Content: "new"},
		{Role: "assistant", Content: "new"},
	}
	// Budget enough for system + last 2 messages but not all
	got := truncateMessages(m, 20)
	if got[0].Role != "system" {
		t.Fatal("first must be system")
	}
	last := got[len(got)-1]
	if last.Content != "new" {
		t.Fatalf("last message should be newest, got %q", last.Content)
	}
}

func TestTruncateAtomicBlock(t *testing.T) {
	m := []llm.Message{
		{Role: "system", Content: "s"},
		{Role: "user", Content: "hi"},
		{Role: "assistant", Content: "", ToolCalls: []llm.ToolCall{
			{ID: "1", Function: llm.FunctionCall{Name: "f", Arguments: "{}"}},
		}},
		{Role: "tool", Content: "result"},
		{Role: "user", Content: "ok"},
	}
	// Enough budget for system + atomic block (assistant+tool) + user
	got := truncateMessages(m, 100)
	if len(got) != 5 {
		t.Fatalf("expected 5, got %d", len(got))
	}
}

func TestTruncateAtomicBlockNotSplit(t *testing.T) {
	tc := llm.ToolCall{ID: "1", Function: llm.FunctionCall{Name: "fn", Arguments: `{"a":"b"}`}}
	m := []llm.Message{
		{Role: "system", Content: "s"},
		{Role: "assistant", Content: "", ToolCalls: []llm.ToolCall{tc}},
		{Role: "tool", Content: "long result that eats budget significantly so we need lots of room"},
		{Role: "user", Content: "x"},
	}
	// Budget enough for user but not the atomic block
	got := truncateMessages(m, 5)
	// Should keep system + user, skip the atomic block entirely
	hasSystem := false
	hasTool := false
	for _, g := range got {
		if g.Role == "system" {
			hasSystem = true
		}
		if g.Role == "tool" {
			hasTool = true
		}
	}
	if !hasSystem {
		t.Fatal("should keep system")
	}
	if hasTool {
		t.Fatal("should NOT include tool without its assistant")
	}
}

func TestEstimateTokens(t *testing.T) {
	msg := llm.Message{Content: "1234567890123456"} // 16 chars = 4 tokens
	got := estimateTokens(&msg)
	if got != 4 {
		t.Fatalf("expected 4, got %d", got)
	}
}

File: sdk/engine/src/toolloop/__tests__/truncate.test.ts
import { describe, it, expect } from 'vitest';
import { truncateMessages } from '../truncate.js';
import type { Message } from '../../types.js';

function msg(role: Message['role'], content: string, toolCalls?: Message['tool_calls']): Message {
  return { role, content, ...(toolCalls ? { tool_calls: toolCalls } : {}) };
}

describe('truncateMessages', () => {
  it('returns all messages when maxTokens is 0', () => {
    const msgs = [msg('system', 'x'), msg('user', 'x'), msg('assistant', 'x')];
    expect(truncateMessages(msgs, 0)).toHaveLength(3);
  });

  it('returns all messages when empty', () => {
    expect(truncateMessages([], 100)).toHaveLength(0);
  });

  it('keeps system message when budget is tight', () => {
    const msgs = [
      msg('system', 'system prompt here'),
      msg('user', 'this is a long user message with lots of text'),
      msg('assistant', 'this is a long assistant reply with lots of text'),
      msg('user', 'another long user message with content'),
      msg('assistant', 'another long assistant reply text'),
    ];
    const result = truncateMessages(msgs, 5);
    expect(result).toHaveLength(1);
    expect(result[0].role).toBe('system');
  });

  it('preserves most recent messages', () => {
    const msgs = [
      msg('system', 's'),
      msg('user', 'old message that is quite long enough to eat budget'),
      msg('assistant', 'old reply that is also quite long to eat budget'),
      msg('user', 'new'),
      msg('assistant', 'new'),
    ];
    const result = truncateMessages(msgs, 20);
    expect(result[0].role).toBe('system');
    expect(result[result.length - 1].content).toBe('new');
  });

  it('keeps atomic blocks together', () => {
    const msgs = [
      msg('system', 's'),
      msg('user', 'hi'),
      msg('assistant', '', [{ id: '1', type: 'function', function: { name: 'f', arguments: '{}' } }]),
      msg('tool', 'result'),
      msg('user', 'ok'),
    ];
    const result = truncateMessages(msgs, 1000);
    expect(result).toHaveLength(5);
  });

  it('does not split atomic block', () => {
    const msgs = [
      msg('system', 's'),
      msg('assistant', '', [{ id: '1', type: 'function', function: { name: 'fn', arguments: '{"a":"b"}' } }]),
      msg('tool', 'this is a very long tool result that takes up significant budget space'),
      msg('user', 'x'),
    ];
    // Budget enough for user but not atomic block
    const result = truncateMessages(msgs, 5);
    const roles = result.map(m => m.role);
    expect(roles).toContain('system');
    expect(roles).not.toContain('tool');
  });

  it('works without system message', () => {
    const msgs = [
      msg('user', 'hi'),
      msg('assistant', 'hello'),
    ];
    const result = truncateMessages(msgs, 1000);
    expect(result).toHaveLength(2);
  });
});
